starlette/responses.py

```diff
@@ -38,7 +38,9 @@ def __init__(
         headers: dict = None,
         media_type: str = None,
         background: BackgroundTask = None,
+        json_encoder: typing.Callable = None
     ) -> None:
+        self.json_encoder = json_encoder
```

Comment:
It feels a little awkward to me to have the option to pass a `json_encoder` to *all* response class `__init__`s.

Is there a reason you wouldn't want to accomplish this pattern by subclassing JSONResponse to override the encoder? (Similar to how UJSONResponse currently works, except overriding the encoding rather than the dumps call)

Reply:
My rational for not subclassing is that would mean a subclass for each encoder type.  Since JSONResponse.render() is just a wrapper over json.dumps it feels slightly better to me to be able to simply pass in the encoder just like I would json.dumps().  


Let me rework this a bit and contain it to JSONResponse and see how it feels.

Reply:
That makes sense to me, but a quick sanity check -- do you need to support lots of different encoder types?

Reply:
OK Here is another approach that confines this to JSONResponse

```
class JSONResponse(Response):
    media_type = "application/json"
    encoder = None

    def render(self, content: typing.Any) -> bytes:
        return json.dumps(
            content,
            ensure_ascii=False,
            allow_nan=False,
            indent=None,
            separators=(",", ":"),
            default=self.encoder
        ).encode("utf-8")
```

We now have to also add 'self' to our encoder method

```
def utf8encoder(self, val):
    if isinstance(val, bytes):
        return val.decode('utf-8')
```

And then we can use it like this...

```
JSONResponse.encoder = utf8encoder
return JSONResponse({'params': params, '__dict__': request.__dict__})
```

Things I like about this approach: 

- Functionality is there
- It's contained to JSONResponse

Things I don't like with this approach.

- The encoder has to have 'self' making it no longer a generic encoder that works with vanilla json.dumps.  Would be nice to have a library of encoders that work with both.
- The encoder could be set at module scope reducing the calling signature for each Response call.

I think the approach that will feel the most naturally as a developer using the framework will be to override the JSONResponse.__init__.  That way it could be fully contained in JSONResponse, but since Response.__init__ has all named parameters this means overriding the entire signature which doesn't seem like a great approach either... 

@dmontagu What's your opinion on the above?  I'd certainly rather have the first approach outlined than not have the ability to pass the encoder at all.  The second approach unless I'm missing something obvious doesn't feel solid either.

Reply:
Note changes are only in comments for now.  If we settle on an approach I'll update PR with code changes.

Reply:
> That makes sense to me, but a quick sanity check -- do you need to support lots of different encoder types? That seems surprising to me, but I might be overlooking something.

I mostly only use str or the utf8encoder method I submitted, bit I like to keep it generic for others.  json.dumps allows us to specify any encoder so I figure why not extend that to JSONResponse?

Reply:
@gnulnx I think there might be a problem here -- because you are using a class variable on JSONRespnose, when you do:
```python
JSONResponse.encoder = utf8encoder
return JSONResponse({'request': request.__dict__})
```
you are actually setting the encoder for *all* JSONResponse instances.

I am not totally sure but, even if you were careful to never use JSONResponse without setting the encoder (which is already an ugly restriction), I could imagine undesired behavior happening if you are changing the encoder type for JSONResponse inside a `run_in_threadpool` call -- it seems to me it's theoretically possible the encoder type could change between where you set it and where the actual dumping happens due to thread safety issues.

The proposed change would work fine though if the approach was to use subclasses of JSONResponse though.

Reply:
I think if you want to have the ability to specify a custom encoder without subclassing, the only reasonable way to do this would be to override the `__init__` and add an `encoder` kwarg:

```python
class JSONResponse(Response):
    def __init__(
        self,
        content: typing.Any = None,
        status_code: int = 200,
        headers: dict = None,
        media_type: str = None,
        background: BackgroundTask = None,
        encoder: Any  # not sure what the type should be
    ):
        super().__init__(content, status_code, headers, media_type, background)
        self.encoder = encoder
```

But you invest much more effort/thought into this, I think it would be good to get confirmation from @tomchristie that this is something starlette would be willing to support (as opposed to preferring the use of subclassing, and keeping the `JSONResponse.__init__` unmodified).

Reply:
I didn't read your post carefully enough -- the adding of self to the encoder seems terrible, but I think there should be a workaround for it... Let me see

Reply:
Using `default=type(self).encoder` drops the need for adding the self argument.

But it's still kind of gross, and either way this encoder-overriding approach will have minor performance downsides.

I personally think both the approach that overrides the `__init__` and the subclassing approach are cleaner/preferable.

Reply:
@dmontagu  I agree that overriding the init method seems cleanest.  Whether that is Response or JSONResponse is I guess really up to @tomchristie.  My initial PR did it in the Response so that we didn't have to duplicate the init signature in both Response and JSONResponse.

Also you are correct about the class variable setting it for all JSONResponse objects.  This could be desirable in many cases.  For instance most projects probably do want to decode as utf8...so maybe you import JSONResponse and then immediately set the encoder to utf8encoder.  Now JSONResponse alway's utf8 encodes for the entire module.  

Personally though, I think it's better to pass it in each time and override init assuming Tom is down with adding this functionality.